Storagebox subaccount

vic1707 opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Adds support for managing storagebox subaccounts.
Fixes #72 (is last PR in TODO)


##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`storagebox_subaccount_info`
##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
I'm not a fan of the API of the module, do you have early suggestions?

felixfontein commented:
Now that I merged the other PR, you need to rebase this one :)

felixfontein commented:
> the `state: present` can either create a subaccount or edit one (makes sense)
> 
>     * if no username is provided the module re-creates the subaccount
>       => So, once ran, the playbook must be updated in order to avoid duplicates

I guess there's no way around, since you cannot provide a username on creation.

Though, actually one could add an option which allows to use `comment` as a unique identifier: if a subaccount with the provided comment exists, update that one (or no change, if already matching options) - otherwise create a new one.

How about having a setting:
`idempotence` with possible values `username` (behavior as now: if username is specified, that user will be edited, otherwise a new user is created), `comment` (user is identified by comment, if there are multiple users with the same comment as supplied fail), with default `username`. (We could potentially add more values in the future.)

>     * if a username is provided we update the subaccount , update the password if provided
>       => no way to know if user wants a random one

How about adding another option `password_mode` with values `update-if-provided`, `ignore-if-exists`, and `set-to-random`? Then the behavior would be:
- If `password_mode=update-if-provided` and `password` is provided, always update the password and report a change.
- If `password_mode=ignore-if-exists`, only use `password` when creating a new subaccount.
- If `password_mode=set-to-random`, always re-set the password to a random one.

(The names are the first ones that came up in my head, I'm sure there are better ones ;) )

I think such an option would be better than having a state `password`, an `ensure_password` boolean flag (you always have to guess that `true` and `false` exactly mean, while the named values are more expressive) or a separate module:

> As noted in the todos, would a state `password` make sense? or a `ensure_password` boolean flag maybe? Should subaccounts password management be a dedicated module?

vic1707 commented:
thx for the advices, I'm still trying to figure out the idempotency thing. 
Just updated a few things, added check mode, a few fixes and more, still not really ready but you might want to take a look again :+1:

Apart from the password changes and username vs comment issues I got 2 new things
1. for check mode, should I add would_<create/edit/edit pass> fields?
2. Right now there's a updating the password when no other modifications are needed results in
```json
{
    "changed": true,
    "failed": false,
    "password_updates": [
        "u462837-sub1"
    ],
    "skipped": [
        "u462837-sub1"
    ]
}
```
if this right in ansible ?

I'll also see how I can make diff mode work

felixfontein commented:
> for check mode, should I add `would_<create/edit/edit pass>` fields?

What would these do? I never heard of anything like that.



> Right now there's a updating the password when no other modifications are needed results in
> 
> 
> ```json
> {
>     "changed": true,
>     "failed": false,
>     "password_updates": [
>         "u462837-sub1"
>     ],
>     "skipped": [
>         "u462837-sub1"
>     ]
> }
> ```
> 
> if this right in ansible ?

I have no idea what `skipped` means. (Also don't name a return value `skipped`, that's a reserved return value by ansible-core for skipped tasks.)

vic1707 commented:
> > for check mode, should I add would_<create/edit/edit pass> fields?
>
> What would these do? I never heard of anything like that.

It was an LLM idea, I didn't find anything related to it either but assumed you'd know way more than me about ansible idioms.

To be honest, I'm learning Ansible and this is my first time writing an ansible module (I made one wrapping a simple cli before but its nothing compared to this collection), and my first time in years writing python, I'm going in pretty much blindly 😓. 
I can write and understand the logic, but the ansible side of thing is as much done with an LLM as it is reading through the docs (which are sometimes confusing, I end up relearning python as much as I'm learning about Ansible modules.

---

Anyway, thanks for the review, I'll fix the issues you're pointing out.
> I have no idea what skipped means. (Also don't name a return value skipped, that's a reserved return value by ansible-core for skipped tasks.)

Didn't know that! It felt like the right name when thinking about the return values 😅 

---

P.S: If you have time, may I ask you if you can write a CONTRIBUTING guide for those like me who would like to help but don't even know how to setup everything (I installed nox, ansible and everything, i think, but can't understand how to properly run the tests nor get proper LSP support within the editor)

vic1707 commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,476 @@
+    required: true
+  subaccounts:
+    description:
+      - The subaccounts configurations.
```
I don't have any opinion on this, I'll let you be the judge !

vic1707 replied:
Nothing in particular, I was simply influenced by using the cloudflare dns one, I find modules handling lists a bit more readable but ansible offers a loop feature so it doesn't change much in the end.

I'll finish my unit tests here, modify the module to take a single subaccount as input, then move on to `password_mode`/`idempotency`

felixfontein replied:
Handling lists is definitely useful if you have a set of resources that you always want to manage at the same time (like all DNS records of the same type for the same name), and I can see some use-cases here as well - but only really if the Robot API would allow creation with specified username.

vic1707 commented on tests/unit/plugins/modules/test_storagebox_subaccount.py:
```diff
@@ -0,0 +1,548 @@
+
+        assert result['msg'] == "Subaccount with username 'u2342-sub1' and homedirectory 'test' has an invalid password (says Hetzner)"
+
+    def invalid_spec_delete_in_username(self, mocker):
```
damn, now I feel even more stupid xD

vic1707 commented:
Finally, I think I'm ready for a final review (and applying your suggestions/requested changes). I kept your original propositions for the `password_mode` values, they're good and didn't feel the need to find better alternatives.

---

One last question, is it a good idea to expose both a `created_subaccount` (containing created subaccount data) and a `created` flag?
Or should `created_subaccount` be the only one present at all time or even be renamed to `created` (as the flag can be easily inferred)? I can see arguments for both so it's hard to decide.

---

Sorry for the thousands notifications you probably got from this PR 😓 and thanks for all the help and advices !

vic1707 commented:
Just realized I'm missing a test for update using comment's idempotency, I'll add it tomorrow

felixfontein commented on plugins/module_utils/robot.py:
```diff
@@ -112,7 +114,11 @@ def raw_plugin_open_url_json(plugin, url, method='GET', timeout=10, data=None, h
             return None, None
-        raise PluginException('Cannot retrieve content from {0}, HTTP status code {1}'.format(url, status))
+        raise PluginException(
+            "Cannot retrieve content from {0} (Error msg: {1}), [{2}] HTTP status code {3}".format(
+                url, reason, method, status
+            )
```
`reason` is basically a stringified version of `status`, so it should be printed next to it. Also `method` belongs to `url`:
```suggestion
            "Cannot retrieve content from {0} {1}, HTTP status code {2} ({3})".format(
                method, url, status, reason
            )
```

felixfontein commented on plugins/module_utils/robot.py:
```diff
@@ -150,7 +156,11 @@ def raw_fetch_url_json(module, url, method='GET', timeout=10, data=None, headers
             return None, None
-        module.fail_json(msg='Cannot retrieve content from {0}, HTTP status code {1}'.format(url, info.get('status')))
+        module.fail_json(
+            msg='Cannot retrieve content from {0} (Error msg: {1}), [{2}] HTTP status code {3}'.format(
+                url, info.get('msg'), method, info.get('status')
+            )
```
Same here:
```suggestion
            msg='Cannot retrieve content from {0} {1}, HTTP status code {2} ({3})'.format(
                method, url, info.get('status'), info.get('msg')
            )
```

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,473 @@
+  password_mode:
+    description:
+      - Controls how password updates are handled.
+      - If C(update-if-provided), the password is updated only if provided (default).
```
I'd rephrase it as follows:
```suggestion
      - If C(update-if-provided), the password is always updated if provided (default).
```

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,473 @@
+      - Controls how password updates are handled.
+      - If C(update-if-provided), the password is updated only if provided (default).
+      - If C(ignore-if-exists), password is only used during creation.
+      - If C(set-to-random), password is reset to a randomly generated one.
```
```suggestion
      - If C(set-to-random), password is reset to a randomly generated one.
      - When a new subaccount is created, the password is set to the specified one if O(password) is provided,
        and a random password is set if O(password) is not provided.
```

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,473 @@
+    choices: [update-if-provided, ignore-if-exists, set-to-random]
+    default: update-if-provided
+    required: false
+  subaccount:
```
I'd remove this suboption and move all its options to top-level. That would be consistent with all other modules in this collection which manage one object.

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,473 @@
+      username:
+        description:
+          - Username of the subaccount.
+          - Required for updating or deleting a subaccount (C(state=absent) or update with C(state=present)).
```
Generally the syntax is `O(option=value)` or `O(option)`, and `C(value)` or `V(value)`.
```suggestion
          - Required for updating or deleting a subaccount (O(state=absent) or update with C(state=present)).
```

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,473 @@
+      password:
+        description:
+          - Password to use or change.
+          - If omitted and state is C(present), a random password is generated by hrobot.
```
```suggestion
          - See O(password_mode) for how and when this is used.
          - Must not be provided if O(password_mode=set-to-random).
```

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,473 @@
+
+      external_reachability:
+        description:
+          - Enable or disable external reachability.
```
```suggestion
          - Enable or disable external reachability (from outside Hetzner's networks).
```

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,473 @@
+      idempotence:
+        description:
+          - Select which attribute to use to check subaccount existence.
+          - C(username) or C(comment).
```
The choices are already listed because of `choices`, so only list them again if there are explanations.
```suggestion
          - If set to C(username), then subaccounts are identified by their username.
            Note that usernames cannot be specified on creation, so you need to use different
            module arguments for creation and updating.
          - If set to C(comment). then subaccounts are identified by their comment.
            If there already exist more than one subaccount with the given comment, the module will fail.
```

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,473 @@
+        required: false
+
+notes:
+  - The module is fully idempotent and supports check mode.
```
The attributes already say that.
```suggestion
```

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,473 @@
+
+notes:
+  - The module is fully idempotent and supports check mode.
+  - Passwords cannot be compared or diffed, so changes to passwords will always appear as changed in check mode and will not show diffs.
```
This is no longer accurate due to `password_mode`. I'd simply remove this.
```suggestion
```

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,473 @@
+notes:
+  - The module is fully idempotent and supports check mode.
+  - Passwords cannot be compared or diffed, so changes to passwords will always appear as changed in check mode and will not show diffs.
+  - When passwords are autogenerated by the API (by omitting the C(password) field), the resulting password is returned.
```
```suggestion
  - When passwords are autogenerated by the API (by omitting the O(password) field), the resulting password is returned.
```

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,473 @@
+created:
+  description: Whether a new subaccount was created.
+  type: bool
+  returned: always
```
Since this isn't returned when the module fails, it should say "success" and not "always".
```suggestion
  returned: success
```
(Same for the other return values.)

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,473 @@
+  type: bool
+  returned: always
+
+created_subaccount:
+  description: The created subaccount object returned from the API.
+  type: raw
+  returned: when created is true
```
Why not simply always return the subaccount data?
```suggestion
subaccount:
  description: The subaccount object returned by the API.
  type: dict
  returned: if O(state=present)
```
I'd also add the `contains:` from `storagebox_subaccount_info.py`'s return value docs.

vic1707 replied:
I really like the idea of returning Hetzner's response. 
But the "always" part might be a bit more difficult:

- subaccounts updates never returns anything
<img width="1616" alt="image" src="https://github.com/user-attachments/assets/01531319-3b80-4757-bf54-e62facc95c6e" />
so we would be completely fabricating the object

- password updates only return the password
<img width="1303" alt="image" src="https://github.com/user-attachments/assets/626fc48a-0357-4e9b-833a-3efff02c8ed8" />

Return raw created account when creating one, for updates/password updates make a new GET request and append the password if provided ?


P.S: the rest of the review comments should already be fixed

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,473 @@
+    )  # pragma: no cover
+
+
+class Subaccount:
```
The collection still supports Python 2.7, so it should say:
```suggestion
class Subaccount(object):
```

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,473 @@
+            if getattr(sa, subaccount["idempotence"], None) == account_identifier
+        ),
+        None,
+    )
```
You need to check that there is at most one subaccount matching the identifier (at least if `module.params["idempotence"] == "comment"`).

vic1707 commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,473 @@
+
+    check_mode = module.check_mode
+    storagebox_id = module.params["storagebox_id"]
+    subaccount = module.params
```
Hetzner's api doesn't really care about extra parameters, but avoiding it was the main reason why subaccount data was in a sub option 🙃

felixfontein commented:
On update, you first do a GET to get the latest subaccount data. You can use that data and update it with the changes that should be applied, and return that (instead of doing another GET after applying the changes).

vic1707 commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,487 @@
+            password_updated = True
+
+        wanted_subaccount = Subaccount(storagebox_id, **subaccount)
+        if wanted_subaccount != existing:
```
The class constructor sets the default values when fields aren't provided, ssh defaults to false (like the hetzner's docs say) etc...
We shouldn't have this problem. Either way I'll add the None filter in the __eq__ implementation

felixfontein replied:
It only sets the default values if the keys aren't part of `subaccount`, but they are part of it. So the values are set to `None` in the `Subaccount` object.

Also just setting them to their default would be wrong as well; not specifying them means "don't change them", not "change them to the default".

vic1707 commented:
Ok, so I return the whole subaccount data at the end, not tied to hetzner's responses 👌
I proposed a new GET request to ensure we have the latest data, I went too far we can absolutely reuse the `existing` variable 🙃

vic1707 commented:
Hi, for the return of the complete subaccount data I had to remove the `Subaccount` class else the code would've gotten way too messy.
Hope everything is good, sorry for the time it took !

---

Do you think a new note should be written about returned `subaccount` prop ?
This prop always contains the full data from hetzner but we might want to emphasize the fact that password is only returned 
- upon creation of the subaccount
- if it got updated (by user specified value or generated by hetzner)
- if nothing happened to it but was provided as input anyway (ie: `password_mode: ignore-if-exists`)

Password is not returned if not provided as input nor updated via `password_mode: set-to-random`

vic1707 commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,456 @@
+
+    if error == "STORAGEBOX_INVALID_PASSWORD":
+        module.fail_json(
+            msg="Creation of user with homedirectory '{0}' has an invalid password (says Hetzner)".format(
```
Old remains of the time where the module was designed to handle multiple subaccounts at once, helped identify them

vic1707 commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,456 @@
+
+    # Contains all subaccount informations
+    # { "subaccount": <data> }
+    return res
```
You're right, it surprisingly doesn't break the tests (it makes sense, but surprising at first), I remembered writing that, don't know why its not here anymore

vic1707 commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,456 @@
+    if error == "STORAGEBOX_INVALID_PASSWORD":
+        module.fail_json(
+            msg="Subaccount with username '{0}' and homedirectory '{1}' has an invalid password (says Hetzner)".format(
+                subaccount["username"] or "<missing>",
```
Actually we can simply return a message saying the password is invalid, we don't need to specify which subaccount we're talking about anymore

vic1707 commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,456 @@
+  username:
+    description:
+      - Username of the subaccount.
+      - Required for updating or deleting a subaccount (O(state=absent) or update with O(state=present)).
```
also required for updates when `idempotence` is `username`.

Basically always required unless we're creating (no way to know beforehand) or when `idempotence=comment`.
Went for 
> Required when using O(idempotence=username) for updates or deletion of a subaccount.

vic1707 commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,456 @@
+  homedirectory:
+    description:
+      - Home directory of the subaccount.
+      - Required when creating or updating a subaccount (O(state=present)).
```
Yes, you're right

> Required only when creating a subaccount (O(state=present)).

vic1707 commented:
you had two comments about the same line, I assume one is for the `username` the other for `comment`,
For username I went for 
> Required when using O(idempotence=username) for updates or deletion of a subaccount.

but kept `required: false` because of the creation usecase.

For comment I went for
> Is required when using O(idempotence=comment) for updates or deletion of a subaccount.

Also keeping `required: false`

I also changed one of the `password` line
```diff
- Must not be provided if O(password_mode=set-to-random).
+ Will be ignored if O(password_mode=set-to-random).
```

vic1707 commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,456 @@
+    description:
+      - Username of the subaccount.
+      - Required for updating or deleting a subaccount (O(state=absent) or update with O(state=present)).
+      - Omit to create a new subaccount (O(state=present)).
```
looks good to me!

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,449 @@
+  diff_mode:
+    support: none
+  idempotent:
+    support: full
```
```suggestion
    support: partial
    details:
      - The Hetzner API does not allow to create subaccounts with specific usernames.
        You can instead use O(comment) to identify accounts by setting O(idempotence=comment),
        that way creation is idempotent.
      - The module is never idempotent if O(password_mode=set-to-random), or if O(password_mode=update-if-provided) and O(password) is specified.
        Set O(password_mode=ignore-if-exists) if you want to provide O(password) on every invocation
        and do not want the module to always change it. Due to how Hetzner's API works, it is not possible
        to query the current password for a subaccount, or check whether a given password is set.
```

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,449 @@
+- name: Create a new subaccount with random password
+  community.hrobot.storagebox_subaccount:
+    storagebox_id: 123456
+    subaccount:
```
You need to adjust the examples, since `subaccount:` no longer exists.

vic1707 replied:
also added a few examples using the `idempotence:comment` and `password_mode:update-if-provided`

felixfontein commented on plugins/modules/storagebox_subaccount.py:
```diff
@@ -0,0 +1,449 @@
+            BASE_URL, storagebox_id, subaccount["username"]
+        ),
+        method="PUT",
+        data=encode_data(subaccount),
```
Here I wouldn't provide the `password` field, to avoid accidental leaks of the password:
```suggestion
        data=encode_data({key: value for key, value in subaccount.items() if key != "password"}),
```

felixfontein reviewed (commented):
I think everything else is ready, there are just a few small things left:

felixfontein commented:
@vic1707 thanks a lot for your contribution!

felixfontein commented:
I'm planning to do a release today or tomorrow.
